### ISSUE TITLE
Add hook to be called when new pending task is available

### ISSUE THREAD

**Issue by Lupus** [smol-rs/async-executor#44]
I'm integrating LocalExecutor into OCaml single-threaded event loop, so far it works great, but what I'm missing is the ability for me to register some non-async thread-safe function that will notify OCaml single-threaded event loop (via a pipe) that it needs to call `LocalExecutor::try_tick`, I have to resort to calling `LocalExecutor::try_tick` "often enough", but that's obviously a suboptimal solution...

Maybe there is something in the API that already allows me to do this without busy-polling?

**Comment by notgull** [comment]
If you wait on `tick()` or `run()` using a custom waker you can integrate `async-executor` pretty well into external loops. Example:

```rust
// Create a waker that wakes up the Ocaml loop when it's ready.
fn wake_up_the_ocaml_loop() { /* ... */ }
let my_waker = waker_fn::waker_fn(|| wake_up_the_ocaml_loop());

// Block on the run() future using the Ocaml loop.
let tick = executor.run(futures_lite::future::pending::<()>());
futures_lite::pin!(tick);

loop {
    // Poll the tick future.
    tick.poll(&mut Context::new(&my_waker));

    // Block on the Ocaml loop until it's awake.
    block_on_ocaml_loop();
}
```

**Comment by Lupus** [comment]
OCaml event loop is the main application that is being executed, Rust executor is auxilary and I want to notify OCaml event loop when there are runnable tasks on the Rust executor, so that OCaml side calls `tick()` to get progress on Rust tasks.

I've ended up with forking `LocalExecutor` from `async-executor`, inlining the multi-threaded executor, removing all the thread-related stuff, which made it really small and it started to work significantly faster on microbenchmark of promise ping-pong between OCaml and Rust. Also the forked version is now capable of running some callback when tasks are added to the queue.

It can be found [here](https://github.com/Lupus/ocaml-lwt-interop/blob/a8d625f/src/local_executor.rs#L129) in `ocaml-lwt-interop` library that allows binding asynchronous Rust to asynchronous OCaml (original copyright notice for `LocalExecutor` is retained of course as this is clearly a derived work).

**Comment by notgull** [comment]
It seems to me from a brief glance that [this](https://github.com/Lupus/ocaml-lwt-interop/blob/a8d625f70e58a832a427a14a70ccbfefbaea6659/src/local_executor.rs#L124) is just a poor man's `Waker`. If you still wanted to use the mainline `async-executor` you could just make a waker that calls whatever you need to use to notify the OCaml loop, then `poll()` the `run()` future with that waker whenever the event source is invoked.

**Comment by valadaptive** [comment]
I'm also running into this issue when trying to tie the executor into egui's event loop.

I have the following function which is run on each iteration of the event loop:

```rust
fn tick(&self) {
    loop {
        let tick = self.executor.tick();
        futures_lite::pin!(tick);
        let poll_result = tick.poll(&mut Context::from_waker(&self.waker));
        if let Poll::Pending = poll_result {
            // No tasks are scheduled. Exit the task loop.
            break;
        }
    }
}
```

`self.waker` is a waker function which wakes up the egui event loop when called.

However, if I spawn a new task onto the executor after all the other tasks have been run, it seems like that task won't be called until the above `tick` function is called by the event loop and sets up its custom waker.

Do I also need to wrap `spawn` to immediately call `tick` after the future is spawned? Is that what you mean by the "event source" being "invoked"? Or am I going about this wrong?

It's starting to feel less like I'm hooking up the executor to an external event loop and more like I'm writing my own half-baked executor on top of yours.

**Comment by notgull** [comment]
You probably want something like this:

```rust
struct EguiExecutorBridge(Box<Pin<dyn Future<Output = ()> + Send + 'static>>);

impl EguiExecutorBridge {
    // Replace with Arc<Executor<'static>>, or however you're storing the executor
    fn new(ex: &'static Executor<'static>) -> Self {
        Self(Box::pin(async move {
            ex.run(futures_lite::future::pending::<()>()).await;
        })
    }

    fn tick(&mut self) {
        // Poll with your egui waker.
        let waker = waker_fn::waker_fn(|| wake_up_my_egui_loop());
        let mut context = Context::new(&waker);
        let _ self.0.as_mut().poll(&mut cx);
    }
}
```

This way, it sets up a local queue and keeps it running. Then, you keep the future around and poll it using the `egui` event loop as a reactor. For more information on how this can be done with GUI event loops, see [`winit-block-on`](https://github.com/notgull/winit-block-on)'s source code.

> However, if I spawn a new task onto the executor after all the other tasks have been run, it seems like that task won't be called until the above `tick` function is called by the event loop and sets up its custom waker.

Yes, you need to call `tick()` or `run()` to drive the executor, especially after a new task has been queued.

> Do I also need to wrap `spawn` to immediately call `tick` after the future is spawned? Is that what you mean by the "event source" being "invoked"? Or am I going about this wrong?

The task being pushed should call the waker and wake up the event loop.

> My mental model--correct me if I'm wrong--is that you run futures by giving them to an executor to schedule. A future needs to be executed using an executor.

Not necessarily. Futures can be blocked on anywhere with [`block_on`](https://docs.rs/futures-lite/latest/futures_lite/future/fn.block_on.html). However, if you want to run a lot of futures, then it's a good idea to use an executor.

**Comment by notgull** [comment]
Closing as I feel that this question has been sufficiently answered.
